[Feat][Core] Implement Event Aggregator Agent

MortalHappiness opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->
This PR implements the event aggregator agent. The event aggregator agent receives events, stores them in a buffer, and then publishes them to external services.

Internal event transmission is done via gRPC, while requests to external services are sent over HTTP. Some Prometheus metrics have also been added. Tested manually (see the image below).

![image](https://github.com/user-attachments/assets/6344a770-7a68-4916-83af-ff5964082535)

## Related issue number

<!-- For example: "Closes #1234" -->
N/A

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

MortalHappiness committed:
[Feat][Core] Implement Aggregator Agent

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Upgrade pytest-httpserver

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness commented on python/ray/dashboard/modules/aggregator/aggregator_agent.py:
```diff
@@ -0,0 +1,251 @@
+)
+
+# Metrics
+if prometheus_client:
```
I didn’t see any other places testing metrics in the codebase. Did you find an example of how it's tested in our codebase?

MortalHappiness replied:
done in https://github.com/ray-project/ray/pull/53182/commits/a749696e2ed6cf028ba0c5db15509cd77c050eea

MengjinYan replied:
Ahh what I meant previously was that, in `test_metrics_agent.py` (https://github.com/ray-project/ray/blob/master/python/ray/tests/test_metrics_agent.py), we've already added tests for all metrics no matter if the metrics was processed by the metrics_agent. And I thought we can add the tests for the event aggregator metrics there as well.

At the same time, I think it is also okay that we add the tests in the aggregator test file as the change now.

MortalHappiness replied:
done in https://github.com/ray-project/ray/pull/53182/commits/12d1fc07612f6aa135c50429efd1336ecb2eb484

MortalHappiness committed:
Address comments

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Address comment: Only wait when event buffer queue is empty

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness commented on python/ray/dashboard/modules/aggregator/aggregator_agent.py:
```diff
@@ -0,0 +1,251 @@
+            except queue.Full:
+                with self._lock:
+                    self._events_dropped_at_receiver_since_last_metrics_update += 1
+                status_code_io_error = 5
```
done

MengjinYan commented on python/ray/dashboard/modules/aggregator/aggregator_agent.py:
```diff
@@ -0,0 +1,251 @@
+            self._events_dropped_at_receiver_since_last_metrics_update = 0
+            self._events_published_since_last_metrics_update = 0
+
+        labels = {
```
Thanks for the context! Looks like "SessionName" can already identify the cluster so no need to add the cluster_id and since it is the common practice for dashboard metrics. Let's just follow that.

MengjinYan commented on python/ray/dashboard/modules/aggregator/aggregator_agent.py:
```diff
@@ -0,0 +1,251 @@
+                self._event_buffer.put(event, block=False)
+                with self._lock:
+                    self._events_received_since_last_metrics_update += 1
+            except queue.Full:
```
This is to follow the current event buffer behavior that we will always keep the latest events in the buffer and drop the oldest events in the buffer when the queue is full. 

And we might not want to keep holding the events in the core worker event buffer which can cause back pressure on the core worker side, especially when the event generation rate is huge.

MortalHappiness committed:
Address comment https://github.com/ray-project/ray/pull/53182#discussion_r2110565176

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Address comment: drop old event instead of new event

https: //github.com/ray-project/ray/pull/53182#discussion_r2110536554
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness commented on python/ray/dashboard/modules/aggregator/aggregator_agent.py:
```diff
@@ -0,0 +1,251 @@
+                    status_code=status_code_unknown_error,
+                    status_message=f"failed to add event to buffer, error: {e}",
+                )
+                return events_event_aggregator_service_pb2.AddEventReply(status=status)
```
done in https://github.com/ray-project/ray/pull/53182/commits/d244894c19d82bae8390ee1ad926a33b7a9375a4

MortalHappiness committed:
Add tests for metrics

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MengjinYan reviewed (commented):
Just one minor comment.

MortalHappiness committed:
Address comment: truncate status message

https: //github.com/ray-project/ray/pull/53182#discussion_r2114476857
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Address comment: Move test_metrics to test_metrics_agent.py

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness commented:
Converted to draft because I am still fixing test errors.

MortalHappiness committed:
Use multiple threads to publish events and handle shutdown gracefully

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Merge branch 'upstream-master' into feature/aggregator-agent

MortalHappiness commented:
@MengjinYan

I did some changes for event publishing in https://github.com/ray-project/ray/pull/53182/commits/4c8226ad2b6b2e64708bbb56f35ba06d5e36085d. Previously, if we used `ThreadPoolExecutor`, only one thread would be selected to run the function regardless of the number of threads in the pool, which I believe is incorrect. I changed it to use `PUBLISH_EVENT_WORKERS` and created that many background threads for publishing events.

For graceful shutdown, I use a `threading.Event` called `self._stop_event` to notify these threads to flush remaining events. I implemented two graceful shutdown paths:

* A SIGTERM handler.
* A background thread `event_aggregator_agent_check_main_thread_liveness` that polls the main thread's liveness.

I manually tested the following cases:

* Sending SIGTERM to `agent.py`: It works.
* Sending SIGTERM to `raylet`: It doesn't work because in `raylet`'s SIGTERM handler, it kills its children by sending SIGKILL.

  * https://github.com/ray-project/ray/blob/ee8fccb94c4751c49c9b09ba1ce63c73ce21d98a/src/ray/raylet/agent_manager.cc#L119
  * https://github.com/ray-project/ray/blob/0617445ecebedf3e9ea5b70a84a0b0be5a89faa0/src/ray/util/process.cc#L569
* Sending SIGKILL to `raylet`: It works because the exit path for `agent.py` is `_check_parent_via_pipe` and it calls `sys.exit(0)`.
  * https://github.com/ray-project/ray/blob/86331a9f2c7c8557ea5fc3da9cc52e6693b43972/python/ray/_private/process_watcher.py#L141

MengjinYan commented on python/ray/dashboard/modules/aggregator/aggregator_agent.py:
```diff
@@ -0,0 +1,304 @@
+                break
+            time.sleep(CHECK_MAIN_THREAD_LIVENESS_INTERVAL_SECONDS)
+
+    def _cleanup(self):
```
It seems that there are multiple places during the graceful shutdown, this `_clean_up` function can be called: 
(1) in `_check_main_thread_liveness`
(2) when receiving a SIGTERM signal

This might 2 cleanups running concurrently in 2 threads. I'm wondering if we should only have one place doing the cleanup to make the shutdown procedure cleaner.

MengjinYan reviewed (commented):
Sorry for the super late review. Added mostly minor comments.

MortalHappiness committed:
Address comments

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Address comment: Add docstrings to all methods

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Address comment: Only run _cleanup once

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Address comments for testing metrics

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MengjinYan commented on python/ray/tests/test_metrics_agent.py:
```diff
@@ -438,6 +449,80 @@ def verify_dashboard_metrics():
+            "ray_event_aggregator_agent_events_published_total": 1.0,
+        }
+        for descriptor, expected_value in metrics_values.items():
+            samples = [m for m in metric_samples if m.name == descriptor]
```
Ahh I guess I meant to verify all the expected labels are set but not necessarily need to verify all their values, if that makes sense.

MortalHappiness committed:
Update python/ray/dashboard/modules/aggregator/aggregator_agent.py

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Add an event for indicating cleanup task done

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Add backoff retry for requests

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness commented on python/ray/dashboard/modules/aggregator/aggregator_agent.py:
```diff
@@ -0,0 +1,363 @@
+            with self._lock:
+                self._events_published_since_last_metrics_update += len(event_batch)
+            event_batch.clear()
+        except Exception as e:
```
Added backoff retry in https://github.com/ray-project/ray/pull/53182/commits/8786ec28aff9f795d0dce41f85a4d73a9e9f88aa

MortalHappiness committed:
Merge branch 'upstream-master' into feature/aggregator-agent

MengjinYan reviewed (approved):
Thank you for following through all the comments! 

I think the retry should eliminate the spamming when the destination doesn't exist. 

Also, a follow up can be to make the default of the destinations to be empty and if they are empty, skip sending to the destination. But it can be put into the later PR to improve the public interface.

MengjinYan commented:
@edoakes or @jjyao Wondering if you can help with merging this one?
